### ISSUE TITLE
update  cache  return . 403 

### ISSUE THREAD

**Issue by vockalimo** [ampproject/amphtml#11455]
Hi all 
   
     I had read the "https://developers.google.com/amp/cache/update-cache" and follow the example  verify the data ok 

> vocka@lalacube:/tmp$ . openssl dgst -sha256 -signature signature2.bin   -verify public-key.pem url.txt .  Verified OK

Step 1. check the public key 
1. https://blog.lalacube.com/.well-known/amphtml/apikey.pub <- can read the data. 

2. my sample code 
`
php
function urlsafe_b64encode($string) {
   $data = base64_encode($string);
    $data = str_replace(array('+','/','='),array('-','_',''),$data);
    return $data;
}
$timestamp=time();
$ampBaseUrl = "https://blog-lalacube-com.cdn.ampproject.org";
//https://blog-lalacube-com.cdn.ampproject.org/c/s/blog.lalacube.com/mei/yahoo_auction_amp.html
$signatureUrl = '/update-cache/c/s/blog.lalacube.com/mei/yahoo_auction_amp.html?amp_action=flush&amp_ts='.$timestamp;
// opening the private key

$pkeyid = openssl_pkey_get_private("file://private-key.pem");
// generating the signature
openssl_sign($signatureUrl, $signature, $pkeyid, OPENSSL_ALGO_SHA256);
openssl_free_key($pkeyid);

$signature = urlsafe_b64encode($signature);
// final url for updating
$ampUrl = $ampBaseUrl.$signatureUrl."&amp_url_signature=".$signature;
echo $ampUrl."\n";
`

example : https://blog-lalacube-com.cdn.ampproject.org/update-cache/c/s/blog.lalacube.com/mei/yahoo_auction_amp.html?amp_action=flush&amp_ts=1506496391&amp_url_signature=hYTPVcaUtEWvg0zPEs-DsDDyeWdjz1RirZLFiFkyeG4-0HFbS7PHBqoREHgZpv60xB3hWXLwTr5b0FHcgOYLIK6_Z48nbMvkY3At7FL0dACIKAaovOxCSigbNJ9xt89gRCYzeg10ub-VM7JDxxazOwuu6lR0UB9BMcRn1-aLcbcX56VSqtmG1c84hpvIIfGnMlpp9O9PRB1Ps55RDRR6pVRrBMldL6FS8YtqQbTFohebL_nPFTLs1_Q5hmQMboPSGmewCcZ4YvtLb6jnw_g60mhU352JV6lyy1QwOmo55WY9QHT9wsC-Co7SnhJJVkBoQDGGBIxY3t8NkaGPnq4QMA

3. step curl 'https://blog-lalacube-com.cdn.ampproject.org/update-cache/c/s/blog.lalacube.com/mei/yahoo_auction_amp.html?amp_action=flush&amp_ts=1506496391&amp_url_signature=hYTPVcaUtEWvg0zPEs-DsDDyeWdjz1RirZLFiFkyeG4-0HFbS7PHBqoREHgZpv60xB3hWXLwTr5b0FHcgOYLIK6_Z48nbMvkY3At7FL0dACIKAaovOxCSigbNJ9xt89gRCYzeg10ub-VM7JDxxazOwuu6lR0UB9BMcRn1-aLcbcX56VSqtmG1c84hpvIIfGnMlpp9O9PRB1Ps55RDRR6pVRrBMldL6FS8YtqQbTFohebL_nPFTLs1_Q5hmQMboPSGmewCcZ4YvtLb6jnw_g60mhU352JV6lyy1QwOmo55WY9QHT9wsC-Co7SnhJJVkBoQDGGBIxY3t8NkaGPnq4QMA'
result is . Error 403 (Forbidden)

I found the https://stackoverflow.com/questions/46176405/amp-update-cache-request-returns-403 , it's  similar. 

So how can i debug ? 
thank 
Vocka

**Comment by advgiarc** [comment]
Having similar issue, with no code, but using this app to test:

https://amp-cache-refresh.appspot.com/

Our public key is publicly available, any clue how to debug this issue would be great.

**Comment by saitho** [comment]
Same here... Update cache gives me a 403: public key is available as plain text, I verified the signature with the public key (SHA256) and server times are synced. I'm using the same code as the issue author above (apart from the additional signature validation).

**Comment by codewiz** [comment]
We're adding some debug info to the 403 responses to make debugging these issues easier in the future. Sorry about that.

I traced the above query and the internal error was "Invalid PEM key". The key decodes correctly with "openssl rsa -pubin -in apikey.pub -text -noout".

However, I noticed that blog.lalacube.com returns a 500 Internal Server Error when I try to fetch the key with curl or wget. The same url works with Chrome. So I suspect this is blocking the AMP Cache from fetching the key.

**Comment by codewiz** [comment]
Looks like the server wants an Accept-Encoding header in the request:

   curl -H 'Accept-Encoding: gzip'  -i https://blog.lalacube.com/.well-known/amphtml/apikey.pub
   HTTP/2 200
   ...

**Comment by advgiarc** [comment]
We're still having the 403 blues, appreciate any help.  Here is our public key: https://minutes.advance.net/.well-known/amphtml/apikey.pub

**Comment by codewiz** [comment]
@advgiarc The PEM key looks correct and the mime-type is text/plain. Please paste a signed update-cache request so I can see where it's being rejected.

Please also provide the code (or pseudocode) you are using to sign the url.

**Comment by advgiarc** [comment]
@codewiz 

https://minutes-advance-net.cdn.ampproject.org/update-cache/c/s/minutes.advance.net/0afe00-rockets-at-knicks-test.amp?amp_action=flush&amp_ts=1507663022&amp_url_signature=JbevOJw3lJwkjIa4ccHJO9H2LKOhfyBRoWT-DmOxZkzLhg-AlIiPeX6H21E6N2W7dOD5OxVmM-vQGy5BIPvjkn_Du4joWT6nQc8_E-RsBrLPkEyo-Bn7-_YIfBehNcmR0t8YxWLGKkTzE78BQ8v4qOtmbHcGQT0c9LnySYNtPt0gMsD0DdNm9ASOQWND-65aUrNb4ix_B_CYMG9UOpfOV3exet08oUrbIGPKJkZ31g7Bkgm0KhWTMyL6fIv_F0IDazBDn5Rr360toSURBaOjBFxSPfDxm16hzb21g3IUcrKos4IqTrSwZEUopCRx-S6BsPw8Ib1sdlQsZSBmMmHe9Q==



`def update_amp_cache(self, amp_url):
    """Use the update-cache signed API request to update the AMP cache
    :param amp_url the url of the amp page
    """
    
    if 'prod' == os.environ.get('tier', 'dev'):
        private_d = self.get_object('amp-key/private-key.pem')
    else:
        with open('ampkeys/private-key.pem') as f:
            private_d = f.read()

    private_key = rsa.PrivateKey.load_pkcs1(private_d, 'PEM')

    amp_ts = int(time.time())
    sign_url = 'update-cache/c/s/minutes.advance.net/{}?amp_action=flush&amp_ts={}'.format(amp_url, amp_ts)
    sig = rsa.sign(message=sign_url, priv_key=private_key, hash='SHA-256')

    if self.verify_sig(signature=sig, message=sign_url):
        amp_cdn_url = 'https://minutes-advance-net.cdn.ampproject.org'
        update_url = '{}/{}&amp_url_signature={}'.format(amp_cdn_url, sign_url, base64.urlsafe_b64encode(sig))

        res = requests.get(update_url)
        self.log.info('AMP cache update-cache result:{} for url: {}'.format(res.status_code, update_url))
        if res.ok:
            return True

    return False


def verify_sig(self, signature, message):
    """verify the signature of the message"""
    # d = self.get_object('.well-known/amphtml/apikey.pub')
    d = requests.get('https://minutes.advance.net/.well-known/amphtml/apikey.pub').content
    public_key = rsa.PublicKey.load_pkcs1_openssl_pem(d)
    verify = rsa.verify(message=message, signature=signature, pub_key=public_key)
    self.log.info('AMP url signature verified? {}'.format(verify))
    if verify:
        return True
    return False
`

**Comment by advgiarc** [comment]
@codewiz 

https://minutes-advance-net.cdn.ampproject.org/update-cache/c/s/minutes.advance.net/f31c00-cleveland-indians-at-new-york-yankees-live-score-updates.amp?amp_action=flush&amp_ts=1507663526&amp_url_signature=aI-OEuC1cXN5MLUqGmN0rU225OWHdd0fYT66ZruV0h8fFGuntZVEQVSnJ_AunplYY9j7Gas76dWNjASDc2ysNipLBaAqD3c5KdMKJOV6YWc2O_9YNI7hA5zrCbtdhGEXZFFQUpjRwUt5lQrK6crlCFaRn5TIgYhebJp3-Jei9mYf3GeHeCMCOGEOFqHZvkX2fMzv_a-MKISjiHByL3odVDGhKx3dihzgrJy0GfPhKl7gMSS3NJw-9iVpWkG4_r5GqMe9At-Lq_ogXFK76cH87N1p77VpbLkMUykuDHyVcOKVebBS3CdPcQldoVVV_TmtRAbkLE7Lk8jtajSTeAbKAg==

pastebin with code:

https://dpaste.de/27OM

Thanks for taking a look!

**Comment by codewiz** [comment]
First of all, I determined that **AMP Cache does not handle HTTP refresh correctly for update-cache verification keys**: if you issue an update-cache request and then swap /.well-known/amphtml/apikey.pub with a different key, we keep using the old key material indefinitely. To make things worse, **404 responses are also cached forever** :-(

I filed an internal bug report, but it might take some time for the fix to roll out to production. Meanwhile, I can flush invalid keys manually. Just send me a private message either on GitHub or on amphtml.slack.com. Apologies for not discovering this sooner.

Specifically for @advgiarc: the code at https://dpaste.de/27OM should include the leading "/" in sign_url. Everything else looks correct to me.

**Comment by advgiarc** [comment]
Great, will post you a private message when back at a computer. Thanks for investigating.

**Comment by vockalimo** [comment]
hi @codewiz 

1.  "curl -H 'Accept-Encoding: gzip' -i https://blog.lalacube.com/.well-known/amphtml/apikey.pub
HTTP/2 200" . <-  Should i enable  all to fetch the apikey.pub ? 

2. I will change  another key , thanks ~

**Comment by vockalimo** [comment]
2. I will change another key , thanks ~ <- cool  .  fixed this . 
thanks

**Comment by jakubszlendak** [comment]
@codewiz Is the issue with cached verification keys solved? Beacause I still get 403 error when trying to flush cache. 
And other question: It is possible to use update-cache request to flush page that has no domain but only IP address (only for testing purposes, eg. https://24-288-88-55.cdn.ampproject.org/c/24.128.88.55/test.html )

**Comment by JoyceBabu** [comment]
I am also getting a 403 error while trying to flush cache

https://gist.github.com/JoyceBabu/5fa508d1ede8f6965601dfce31d1cc8f

**Comment by yumruk** [comment]
hello,

How can I fix?
**404 error:**

> https://habervaka-com.cdn.ampproject.org/r/s/habervaka.com/.well-known/amphtml/apikey.pub 

**403 error:**

> https://www-habervaka-com.cdn.ampproject.org/update-cache/c/s/www.habervaka.com/sureli/mehmet-demirkol-galatasaraya-gelebilir-07-january-2018.html/amp?amp_action=flush&amp_ts=1516171261&amp_url_signature=AlIznL0gKFOjOv5btv8buZhpc5W46uv6zw3YUegrFOx0Q_NdkE-EST7k0gYCYrI5JxEhoti-bWZOMlya9-Uocgv5n9xCSkE4ww5iVSFjvj46uZUtE54ExduxZJy8l4rhvGvlv9_dapcM9Mgb9id9lv7E58_cV3jjfNhduKsZ54EIuEq9w8oTEFmsujUfA1vVmpw3R2q7lk9A1-gPmhqS4tEKcWWODxHIZ5esBDkUzOMynSNXJNsXzJZPx5pK8SlESRcd72Nh-veb-xm8mTgCXrDUEwESjkuAuQxh9hs3nu1yy-uEZBVf-7cYyKj_etilOncmqZ7KTdTDFY6Cxp659A

**Comment by saitho** [comment]
@yumruk Your website's (https://habervaka.com) SSL certificate was revoked. (SEC_ERROR_REVOKED_CERTIFICATE). This might be the issue. ;)

**Comment by yumruk** [comment]
@saitho To fix the problem I'm talking about ı have again uploaded an new SSL certificate. This is the error you mentioned.
How can I fix?
404 error:

https://habervaka-com.cdn.ampproject.org/r/s/habervaka.com/.well-known/amphtml/apikey.pub

403 error:

https://www-habervaka-com.cdn.ampproject.org/update-cache/c/s/www.habervaka.com/sureli/mehmet-demirkol-galatasaraya-gelebilir-07-january-2018.html/amp?amp_action=flush&amp_ts=1516171261&amp_url_signature=AlIznL0gKFOjOv5btv8buZhpc5W46uv6zw3YUegrFOx0Q_NdkE-EST7k0gYCYrI5JxEhoti-bWZOMlya9-Uocgv5n9xCSkE4ww5iVSFjvj46uZUtE54ExduxZJy8l4rhvGvlv9_dapcM9Mgb9id9lv7E58_cV3jjfNhduKsZ54EIuEq9w8oTEFmsujUfA1vVmpw3R2q7lk9A1-gPmhqS4tEKcWWODxHIZ5esBDkUzOMynSNXJNsXzJZPx5pK8SlESRcd72Nh-veb-xm8mTgCXrDUEwESjkuAuQxh9hs3nu1yy-uEZBVf-7cYyKj_etilOncmqZ7KTdTDFY6Cxp659A

**Comment by codewiz** [comment]
@yumruk The 403 error page says: "Invalid PEM key."

The key looks good to me, but there's no Content-Type header in the response. Please configure your webserver to serve the key as "text/plain".

**Comment by yumruk** [comment]
@codewiz  I did "openssl rsa -pubin -in apikey.pub -text -noout". Do I do the right thing?

**Comment by rnagaraj** [comment]
Hi @codewiz 
I was trying to implement update cache, encountered few issues and have few questions.
1. Do we really have to host the public key at the location **.well-known/amphtml/apikey.pub** ?
2. The site where we are implementing ampcache is secured. Do we have to do any configuration to work over **https**

when I tried updating the cache , I got 403 and the response as below (Public Key is still not made public)
`HttpResponseProxy {
	HTTP / 1.1 403 Forbidden[Content - Type: text / html;
		charset = UTF - 8, X - Content - Type - Options: nosniff, Date: Tue, 23 Jan 2018 20: 09: 27 GMT, Server: sffe, Cache - Control: private, X - XSS - Protection: 1;
		mode = block, Alt - Svc: hq = ":443";
		ma = 2592000;
		quic = 51303431;
		quic = 51303339;
		quic = 51303338;
		quic = 51303337;
		quic = 51303335, quic = ":443";
		ma = 2592000;
		v = "41,39,38,37,35", Transfer - Encoding: chunked]
}`

**Comment by advgiarc** [comment]
@rnagaraj 
(I am not associated with the ampproject) 

We are using AMP and I can offer some insight to your questions:

1) Yes, that is the exact location your public key must live under.

2) No, if your SSL certs are all setup correctly, there is nothing else you need to do for HTTPS, as AMP and 'update-cache' ONLY work over HTTPS.

RE: the 403: if your public key is not actually public and at the .wellknown location, things will not work.

**Comment by rnagaraj** [comment]
Thank you @advgiarc

**Comment by fdiego26** [comment]
I have the same 403 error Invalid PEM key
https://kamisetas.com/.well-known/amphtml/apikey.pub

403 error
https://kamisetas-com.cdn.ampproject.org/update-cache/c/s/kamisetas.com/amp/parches-bordados/Homer-en-el-seto-parche.html?amp_action=flush&amp_ts=1519225435&amp_url_signature=j_VAs7hxAxKweXsX8eQqynP6IJZG4MCgpcYob6IJgOejzPU_3NOwC0G8jXgFGxDe1ZHZ6kwGIME9ydzfnATUP8Y13lQ6ErXMP0uu1yB4DGzKGN8ntY72EA54sLM5RipFOMtZUENwJtuyQz2eohtIbAfudI51fXob4N7kYG8rDtVFSJl1zOsHF8lchIDoBY8aI8zgZXDhjtOEWl7BGGQfeRfPIB-4brOBSglXXxtveDS1ZuhnuVjrnHMTV34To8OZ1i16b_qNRmF8TSls1jRAB8us6mFugoOfMMVVfmP9LOCbJVg2lkh02zsTUzTLuiwTJawNJB9rEMUX_hfpisjlHQ

Thanks, from Spain

**Comment by ogun** [comment]
Hello all,

I'm still getting 403 "Invalid PEM key" error message but I couldn't find any problem. Could you please control our implementation?

https://gist.github.com/ogun/61d8396eb6c58ab3c5ef71d61df89bad

**Comment by flo-rider** [comment]
@codewiz Hi! Could you help me flush my apikey on toptrouwen.nl/.well-known/amphtml/apikey.pub from Google AMP cache?

**Comment by Dimasmagadan** [comment]
Hello @codewiz 
same issue. Could you please help flushing my API key?
https://studentloanhero-com.cdn.ampproject.org/r/s/studentloanhero.com/.well-known/amphtml/apikey.pub

I keep getting URL signature verification error.

Attached is generated link:
/update-cache/c/s/studentloanhero.com/featured/fico-auto-score-guide/?amp_action=flush&amp_ts=1523305916&amp_url_signature=aNi4NwXpQJwrLUCqWi-_z9WreySJ19gSvOOR8IC8GcYL2vfzDZiSxtPsTiwRb5I1qRNZrKO_EFhs0eYilM2duXZNqf__uYBvIzAlTpCWT6VESutomI1mbtIKa_1OsJPWWZPF_Madm-ma6aZQdrR2CvUXnWHqmYMve0wfwO-6BsHx0qux3jpsWgGA5t3al6UVWjOKAcaURUJYI8l_ZII_cX_is3EOrW7la3Qhel8-Kto3eEKnny2Up5VivvfhQKufDfe74k91SRR2wys26aPRr7ZsDq8pLCm5CK3f0LYFvplbx5s-OaT-8muL99alZb6z9HBX-KJt2FPES6bdEoh6Jw==

I've checked, it verifies ok with openssl locally.

**Comment by obeyer** [comment]
Hi @codewiz,
Could you please help with the same issue that other have? I keep getting:
403. That’s an error.
Your client does not have permission to get URL /update-cache/c/s/www.popsugar.com/fashion/Olivia-Munn-Red-One-Piece-Swimsuit-44352095/amp?amp_action=flush&amp_ts=1523601115&amp_url_signature=AUNLk1uRInwqbvJm11wPnHAhRDbmb68lqIpdBWJZ5-xRfwRZQfTLNR9zqQ5fbii0MkHwCh9z0U--0yJCewMc1kZ_S1Z1uLt7WJBxIyLw4Jjqvu53i-hRl4RyNklPLC3gYvA0NWHtKkoD8G6eI8TQ6aN9Le-LKA5NRQ4HkWAraLNs2JtlKy2bCOqHzLAwMUe9z5FnPkFcdsAZmqft_Prf3nEDwXXbzml8fUY_NMn33eW45tn3Ss_6EfSQHJ1ALtU2xuhf61awurRsMMbvfec469GDybRiSNyZnBn-h_Y0ZV0xvunL6C1smTydd9jA11GRmOFsArSnV69Cm73dc1iXpA== from this server. (Client IP address: 2601:645:c480:bc0:5ce4:2d2c:c66c:bc40)

URL signature verification error. That’s all we know.

**Comment by obeyer** [comment]
I had the same issue. my problem was:
1) I used phpseclib\Crypt\RSA, and it wasn't signing it in a way that was expected apparently. once I switched to openssl php library, and used OPENSSL_ALGO_SHA256 in openssl_sign(), it finally worked! 2) also, I was only replacing + and / for web-safe encoding. After reading comments, I added = to it as well, so it looks like this:
$encoded_signature = str_replace(['+','/','='],['-','_',''], base64_encode($signature));
3) I made sure that I have a leading / in the url request that I sign (the one before update-cache).

Hope that helps!

**Comment by Dimasmagadan** [comment]
Hey @obeyer 
thanks for an update. I've tried to replace '=' as well, but it didn't work.

The resulting generated link is this one:
/update-cache/c/s/studentloanhero.com/featured/fico-auto-score-guide/?amp_action=flush&amp_ts=1523607919&amp_url_signature=pj_Q-MhePQ2_vA9i-MTe4pBva5dy7FiXbusIBRY4yM0I9csgFG_-oZDxQcB7nx1RgTmoQw6reNJnDzzTKr1rQeK_UnT1e0E7UtshSkK545eMv3wV-PTDiWc4d2L8DW7FbkqoHUsT_gDEMdbuppnGLf7qpK2YTlIAb7_H8iQvvXr6rOaxyY_vE2lQ7nANov8dotjRR8Ao0ZenCwffGRasAnBLCwPmFhhw1aMfmFy6NLOQyXM-Tlv2Au2fhehI0I7TTgdroGW-8JjDbROucxXLnw58pfUWsJYijxnPV37dTPXxH6XgJwTi3Zro9hcT73q0oKOvQ7xRl_GPgz5BFwWQug

**Comment by Arijit1303** [comment]
Hello 
Iam having an issue with the amp_cache update , I followed the guide provided it this link 
-------------------------------------------------------------------
 https://developers.google.com/amp/cache/update-cache 
-------------------------------------------------------------------
which is not very specific and a bit hard to follow 
My Apikey link is =  https://everest.trodly.com/.well-known/amphtml/apikey.pub
---------------------------------------------------------------------------------------------
I have used PHP as backend 
used base_64encode() to encode the signature 
following is my URL :
---------------------------------------------------------------------------------------------
https://everest-trodly-com.cdn.ampproject.org/update-cache/c/s/everest.trodly.com/blog/article/when-i-discovered-my-love-for-the-mountains/amp?amp_action=flush&_ts=1524569136&_url_signature=fnD8W9Ri19yINaABdM0pT5rxn9eJ5XXwNu6SSdAnfpNWU4lGdfK3v1WWrn9CvKs+LTG3vrsMtgaVcv3ALecSRxxJy5BI+7wocW35bdZO4n/sXXiqOlOERdP6ifAJNZgHsitbiu6j3Vw86lzUSAVhGVDMe0nAYLjneaZOBaogDX46zeEfE7T7ho9T215nFSjoTCTbMZavLMOqK8oKODz21kvBnd2wUY9rj9P2HfhoP9k6wiNDi1ads6CKWjxQ0+hTxFjN9+jAm6HBB83j6xsD+p1i8ofwEyq2BXnjlkg1a5dIiQzlDg4yTkRrJTDuxw3/f6H3CpsdzQor2hpOxWdJEg==
---------------------------------------------------------------------------------------------
This Verified Ok locally , 
But when I used curl to execute this is what it returns 

"404 not found : Unable to decode Signature"
----------------------------------------------------------------------------------------------

Can anyone help me ?

**Comment by JoyceBabu** [comment]
I have published a composer package for clearing AMP cache

https://packagist.org/packages/ennexa/amp-update-cache

**Comment by obeyer** [comment]
@Dimasmagadan your key gets downloaded, if I click on the link. it should just return the key, like this: http://www.popsugar.com/.well-known/amphtml/apikey.pub

**Comment by csLittleye** [comment]
Have you successfully updated the cache before?

Did you ever changed your public keys before you got the errors?

**Comment by manman1992** [comment]
@codewiz same issue. Could you please help flushing my API key?
[https://www-lalamira-com.cdn.ampproject.org/r/s/www.lalamira.com/.well-known/amphtml/apikey.pub](https://www-lalamira-com.cdn.ampproject.org/r/s/www.lalamira.com/.well-known/amphtml/apikey.pub)

**Comment by manman1992** [comment]
@codewiz 

GET https://www-lalamira-com.cdn.ampproject.org/update-cache/c/s/www.lalamira.com/amp/Cheap-Bridesmaid-Dresses-c7?amp_action=flush&amp_ts=1535099839&amp_url_signature=E-a3ZOSJjxwqeNt-zFBeaXcG5BSvAV8BUvFdGKieURxhN7SVQzJGG4MJZ3ZxZ7jDCX03N36uWFysKkw8eQIQtxU8_yu0xtA49QxvmKDBgJTaGjZQauR8IVJ7mb4GDaOYg5Av3VYOKweJuE8vZHQl5gOa6UxDMYNxUGL_RBu4xVLnASd-l0ZwAcLJnvhsVAqHqjrChlJPhQkbLlGxJFjJdQEwsevT4p-tpKLNggEiCL8S0VL0_JI8dYjJvVNQIuwyekU4x8w26nIxoziXI9xCSZpWEoNqXt7mZxmgj_9leqfA4sQPpjX3yZFS-9gHSboa1BRpVE91VYl_9rioEKLHBQ` resulted in a `403 Forbidden`

help me  thank you !

**Comment by manman1992** [comment]
solved thank you

**Comment by jorakaraivanov** [comment]
Hello @codewiz

Could you flush my next keys:

1. https://www-tv2nord-dk.cdn.ampproject.org/r/s/www.tv2nord.dk/.well-known/amphtml/apikey.pub
2. https://www-tvmidtvest-dk.cdn.ampproject.org/r/s/www.tvmidtvest.dk/.well-known/amphtml/apikey.pub
3. https://www-tveast-dk.cdn.ampproject.org/r/s/www.tveast.dk/.well-known/amphtml/apikey.pub
4. https://www-tvsyd-dk.cdn.ampproject.org/r/s/www.tvsyd.dk/.well-known/amphtml/apikey.pub
5. https://www-tv2fyn-dk.cdn.ampproject.org/r/s/www.tv2fyn.dk/.well-known/amphtml/apikey.pub
6 .https://www-tv2ostjylland-dk.cdn.ampproject.org/r/s/www.tv2ostjylland.dk/.well-known/amphtml/apikey.pub

**Comment by fueledbymarvin** [comment]
@codewiz Would really appreciate it if you helped me out by flushing the key here: https://genius-com.cdn.ampproject.org/r/s/genius.com/.well-known/amphtml/apikey.pub.

**Comment by Gregable** [comment]
Reported recent updates internally to the AMP Cache product. Internal Id 117440746

**Comment by fueledbymarvin** [comment]
@Gregable do you know when we can expect to see our key's cache refreshed? We've been experiencing problems for many weeks now.

**Comment by csLittleye** [comment]
To have your key refreshed: https://developers.google.com/amp/cache/update-cache#update-rsa-keys

**Comment by csLittleye** [comment]
@fueledbymarvin, particularly for your key, it was already refreshed on 2018/10/11.

**Comment by fueledbymarvin** [comment]
@csLittleye Thank you! It appears to be working now.

**Comment by valuDe** [comment]
Is the key caching still an issue?

Our public key is on https://cartop-carport.de/.well-known/amphtml/apikey.pub

I tried refreshing it multiple times using the explanation provided on
https://developers.google.com/amp/cache/update-cache#update-rsa-keys
but 
https://cartop--carport-de.cdn.ampproject.org/r/s/cartop-carport.de/.well-known/amphtml/apikey.pub
always gives me 404. I can't spot the error.

For
https://cartop--carport-de.cdn.ampproject.org/update-cache/c/s/cartop-carport.de/amp/?amp_action=flush&amp_ts=1571060204&amp_url_signature=EjSUVz9fe4JH7W67p8G7Hy8hXd79skon6g-d4Y08u0X7vzqV5lop_6T-gICc1kpnzsE4VvNSRp5F3Aow5AJRSrjkwveP9wwMV0iKDTjvP24cfF4F4T5RfqWL_3AWz1pdCq4608JWWzhIjCxyF2JBzMTeCjx_NQ6G2DzMe0R8ZhkFRrnZzE4h7QdDLpkL8zs63g6rcb6jHIH0foye9DGHC2QNsOpwEp5ZTzFsVUnGRoyLAxTi2KGwMeauDi_4yALiykdbzdBoo0zQsW8yz8L0gdUx-LTFG-OlJMwc4k1yv7FWx937IQvO_E8aCxGfswiN0bFBawU9xOLG_AcxP6JXNg==
we get
_Invalid public key due to ingestion error: Invalid Content_

@codewiz if caching is still an issue, could you flush our public key?

**Comment by GuysAtILSS** [comment]
Hey @codewiz can you please flush our apikey.pub too?
https://m-penoblo-de.cdn.ampproject.org/r/s/m.penoblo.de/.well-known/amphtml/apikey.pub

very thank you!

**Comment by GuysAtILSS** [comment]
Is there any chance we can solve our issue (one post before) ?
I really don't know who else can help us with this problem.

Thank you!

**Comment by acidre** [comment]
Hello there,

The generated URL returns a "OK" response, but the AMP cache is not updated, is there something I might miss?

Thanks,

**Comment by fmsiutti** [comment]
Hey @codewiz, I don't know if this is still going, but i can't access my pub key from amp cdn.
Can you please flush this too?
https://amp-pisamonas-es.cdn.ampproject.org/r/s/amp.pisamonas.es/.well-known/amphtml/apikey.pub

Thank you!

**Comment by conroywhitney** [comment]
@obeyer -- Thank you! Your comment was the one that finally got me across the finish line. It works! 🎉 

For those who are still struggling with it, here's my Ruby on Rails code (I just put it in `app/models/amp_purge.rb` to test):

https://gist.github.com/conroywhitney/5084c51148bd2dd2a00e6c95947758ac

Call it like `AmpPurge.new("/foo/bar.amp").purge`, then open the final "Signed URL" in your browser, you should get "OK"

I was messing up a whole bunch of stuff, so here's what I learned:

 1. You need to sign the whole path, but not the full URL, so this: `/update-cache/c/s/www.example.com/cooking/best-right-now/the-best-cookware-sets.amp?amp_action=flush&amp_ts=1613839137`

 2. It looks like the the trailing `==`s at the end of the Base64 signature don't really matter, but do make sure that you're using the web-safe base64 version that doesn't include `+` or `/` (https://ruby-doc.org/stdlib-2.4.0/libdoc/base64/rdoc/Base64.html#method-i-urlsafe_encode64)

 3. Make sure you're using the AMP URL that gets resolved, so like in my case, `https://example-com.cdn.ampproject.org/c/s/example.com/foo/bar.amp` was actually redirecting to `https://www-example-com.cdn.ampproject.org/c/s/www-example.com/foo/bar.amp`. I didn't realize our AMP name included the `www` so I needed to make sure that was part of my `update-cache` request.

 4. Use the `.pem` key that has the `-----BEGIN RSA PRIVATE KEY-----` to sign your requests. These two articles were helpful in understanding how to sign URLs in Ruby: https://docs.ruby-lang.org/en/2.1.0/OpenSSL.html#module-OpenSSL-label-Loading+a+Key and https://tomkadwill.com/message-signing-with-ruby-openssl-rsa

 5. Like other people have pointed out, make sure your public key is correct:
  a. It needs to start with `-----BEGIN PUBLIC KEY-----`
  b. It needs to be hosted at `https://www.example.com/.well-known/amphtml/apikey.pub`
  c. It needs to be served as `text/plain`
  d. It needs to be excluded from `robots.txt`
  e. It's in the AMP cache correctly at `https://www-example-com.cdn.ampproject.org/r/s/www.example.com/.well-known/amphtml/apikey.pub`

That's about all I got. Best of luck!

**Comment by davidfloyd91** [comment]
> I have published a composer package for clearing AMP cache
> 
> https://packagist.org/packages/ennexa/amp-update-cache

this worked for me. got it up and running relatively quickly despite never having used php
